deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/java/org/deeplearning4j/iterator/BertIterator.java

```diff
@@ -0,0 +1,479 @@
+/*******************************************************************************
+ * Copyright (c) 2015-2019 Skymind, Inc.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+package org.deeplearning4j.iterator;
+
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import org.deeplearning4j.iterator.bert.BertMaskedLMMasker;
+import org.deeplearning4j.iterator.bert.BertSequenceMasker;
+import org.deeplearning4j.text.tokenization.tokenizer.Tokenizer;
+import org.deeplearning4j.text.tokenization.tokenizerfactory.BertWordPieceTokenizerFactory;
+import org.deeplearning4j.text.tokenization.tokenizerfactory.TokenizerFactory;
+import org.nd4j.base.Preconditions;
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.dataset.api.MultiDataSet;
+import org.nd4j.linalg.dataset.api.MultiDataSetPreProcessor;
+import org.nd4j.linalg.dataset.api.iterator.MultiDataSetIterator;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+import org.nd4j.linalg.primitives.Pair;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * BertIterator is a MultiDataSetIterator for training BERT (Transformer) models in the following way:<br>
+ * (a) Unsupervised - Masked language model task (no sentence matching task is implemented thus far)<br>
+ * (b) Supervised - For sequence classification (i.e., 1 label per sequence, typically used for fine tuning)<br>
+ * The task can be specified using {@link Task}.
+ * <br>
+ * Example for unsupervised training:<br>
+ * <pre>
+ * {@code
+ *
+ * }
+ * </pre>
+ *
+ * This iterator supports numerous ways of configuring the behaviour with respect to the sequence lengths and data layout.<br>
+ * <br>
+ * <u><b>{@link LengthHandling} configuration:</b></u><br>
+ * Determines how to handle variable-length sequence situations.<br>
+ * <b>FIXED_LENGTH</b>: Always trim longer sequences to the specified length, and always pad shorter sequences to the specified length.<br>
+ * <b>ANY_LENGTH</b>: Output length is determined by the length of the longest sequence in the minibatch<br>
+ * <b>CLIP_ONLY</b>: For any sequences longer than the specified maximum, clip them. If the maximum sequence length in
+ * a minibatch is shorter than the specified maximum, no padding will occur.<br>
```

Comment:
What happens in the ANY_LENGTH and CLIP_ONLY for the shorter sequences? Padding?

Reply:
We pad only to the longest sequence in the minibatch.